src/state_oorset_ps.erl

```diff
@@ -0,0 +1,890 @@
+equal({?TYPE, {DataStoreA, FilteredOutEventsA, {vclock, AllEventsA}}},
+      {?TYPE, {DataStoreB, FilteredOutEventsB, {vclock, AllEventsB}}}) ->
+    DataStoreA == DataStoreB andalso
+        FilteredOutEventsA == FilteredOutEventsB andalso
+        lists:sort(AllEventsA) == lists:sort(AllEventsB).
+
+%% @doc Check if an ORSet is bottom.
+-spec is_bottom(delta_or_state()) -> boolean().
+is_bottom({?TYPE, {delta, ORSet}}) ->
+    ORSet == {{orddict:new(), orddict:new()}, ordsets:new(), ordsets:new()};
+is_bottom({?TYPE, _ORSet}=FullORSet) ->
+    FullORSet == new().
+
+%% @doc Given two `state_oorset_ps()', check if the second is and inflation of the first.
+%% The inflation will be checked by the is_lattice_inflation() in the common library.
+-spec is_inflation(state_oorset_ps(), state_oorset_ps()) -> boolean().
+is_inflation({?TYPE, ORSet1}, {?TYPE, ORSet2}) ->
+    is_inflation_oorset_ps(ORSet1, ORSet2).
+
+%% @doc Check for strict inflation.
+-spec is_strict_inflation(state_oorset_ps(), state_oorset_ps()) -> boolean().
+is_strict_inflation({?TYPE, _}=CRDT1, {?TYPE, _}=CRDT2) ->
+    state_type:is_strict_inflation(CRDT1, CRDT2).
+
+%% @todo
+%% @doc Join decomposition for `state_oorset_ps()'.
+-spec join_decomposition(state_oorset_ps()) -> [state_oorset_ps()].
+join_decomposition({?TYPE, ORSet}) ->
+    [ORSet].
+
+%% @private
+remove_all_private([], _ElemDataStore, ResultEvents) ->
+    ResultEvents;
+remove_all_private([Elem|RestElems], ElemDataStore, ResultEvents) ->
+    case orddict:find(Elem, ElemDataStore) of
+        {ok, Provenance} ->
+            ElemEvents = get_events_from_provenance(Provenance),
+            remove_all_private(RestElems,
+                               ElemDataStore,
+                               ordsets:union(ResultEvents, ElemEvents));
+        error ->
+            {error, {precondition, {not_present, Elem}}}
+    end.
+
+%% @doc @todo
+-spec subtract_all_events(ps_all_events(), ps_dot()) -> ps_dot().
+subtract_all_events({vclock, AllEvents}, Events) ->
+    AllEventsSet = lists:foldl(
+                     fun({EventId0, Counter0}, AllEventsSet0) ->
+                             get_ordsets_from_vclock({EventId0, Counter0},
+                                                     AllEventsSet0)
+                     end, ordsets:new(), AllEvents),
+    ordsets:subtract(AllEventsSet, Events);
+subtract_all_events(AllEvents, Events) ->
+    ordsets:subtract(AllEvents, Events).
+
+%% @doc @todo
+-spec add_elem_with_dot(element(), ps_dot(), ps_data_store()) -> ps_data_store().
+add_elem_with_dot(Elem, Dot, {ElemDataStore, EventDataStore}) ->
+    NewElemDataStore = add_elem_with_dot_private(Elem, Dot, ElemDataStore),
+    NewEventDataStore =
+        ordsets:fold(
+          fun(Event, NewEventDataStore0) ->
+                  add_elem_with_dot_private(Event, Elem, NewEventDataStore0)
+          end, EventDataStore, Dot),
+    {NewElemDataStore, NewEventDataStore}.
+
+%% @doc @todo
```

Comment:
For merging definitely I have to finish it, but I am not sure about reviewing. If you want to review this with documentation and commenting, I might need a couple of days to completing them.